base/kra/src/main/java/org/dogtagpki/server/kra/rest/base/KeyRequestProcessor.java

```diff
@@ -0,0 +1,1050 @@
+                    "certServer.kra.request", "read");
+            return createKeyRequestInfo(request, baseUrl);
+        } catch (EAuthzAccessDenied e) {
+            throw new UnauthorizedException("Not authorized to get request");
+        } catch (EBaseException e) {
+            throw new PKIException(e.getMessage(), e);
+        }
+    }
+
+    public KeyRequestResponse submitRequest(Principal userPrincipal, String baseUrl, RESTMessage data) {
+        Object request = null;
+
+        try {
+            Class<?> requestClazz = Class.forName(data.getClassName());
+            request = requestClazz.getDeclaredConstructor(RESTMessage.class).newInstance(data);
+        } catch (ClassNotFoundException | NoSuchMethodException | SecurityException | InstantiationException
+                | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
+            throw new BadRequestException("Invalid request class." + e, e);
+        }
+
+        logger.debug("KeyRequestProcessor: Request class: {}", request.getClass().getSimpleName());
+        if (request instanceof KeyArchivalRequest) {
+            return archiveKey(userPrincipal, baseUrl, new KeyArchivalRequest(data));
+
+        } else if (request instanceof KeyRecoveryRequest) {
+            return recoverKey(userPrincipal, baseUrl, new KeyRecoveryRequest(data));
+
+        } else if (request instanceof SymKeyGenerationRequest) {
+            return generateSymKey(userPrincipal, baseUrl, new SymKeyGenerationRequest(data));
+
+        } else if (request instanceof AsymKeyGenerationRequest) {
+            return generateASymKey(userPrincipal, baseUrl, new AsymKeyGenerationRequest(data));
+
+        } else {
+            throw new BadRequestException("Invalid request class.");
+        }
+
+    }
+
+    public void approve(Principal userPrincipal, RequestId id) {
+        if (id == null) {
+            throw new BadRequestException("Invalid request id.");
+        }
+        if (userPrincipal == null || userPrincipal.getName() == null) {
+            throw new UnauthorizedException("Request approval must be initiated by an agent");
+        }
+        try {
+            Request request = requestRepository.readRequest(id);
+            authz.checkRealm(request.getRealm(), getAuthToken(userPrincipal),
+                    request.getExtDataInString(Request.ATTR_REQUEST_OWNER),
+                    "certServer.kra.requests", "execute");
+
+            kra.addAgentAsyncKeyRecovery(id.toString(), userPrincipal.getName());
+
+            auditor.log(new SecurityDataRecoveryStateChangeEvent(
+                    userPrincipal.getName(),
+                    ILogger.SUCCESS,
+                    id,
+                    "approve"));
+
+        } catch (EAuthzAccessDenied e) {
+            throw new UnauthorizedException("Not authorized to approve request", e);
```

Comment:
This is identical to the original code, but do you think we should log an audit event here? No need to fix the code now, but maybe just add a TODO comment for a future review.

Reply:
In this exception (and some other) there is no audit failure while in other there is. I have added the TODO in all cases for future evaluation.

In my opinion this should not generate a failure event for the operation in the audit because the authorisation stop any processing before it start. An authorisation failure event could be possible but not one related to the following operation. 
An improvement of this would be to move the authorisation check in the ACL filter and decide what the audit should include for the authorisation event.